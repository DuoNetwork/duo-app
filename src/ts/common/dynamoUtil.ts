import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
