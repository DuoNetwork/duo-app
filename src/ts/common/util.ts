import { IContractPrice } from '@finbook/duo-contract-wrapper';
import { IPriceStatus, IStatus } from '@finbook/duo-market-data';
import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import { ISourceData } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public formatTime(e: number) {
		return moment(e).format('YYYY-MM-DD HH:mm:ss');
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<IContractPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gemini: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gdax: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
