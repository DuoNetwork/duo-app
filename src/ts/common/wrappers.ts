import {
	Constants as WrapperConstants,
	DualClassWrapper,
	EsplanadeWrapper,
	ICustodianWrappers,
	MagiWrapper,
	StakeWrapper,
	Web3Wrapper
} from '@finbook/duo-contract-wrapper';

export const web3Wrapper = new Web3Wrapper(
	window,
	(__KOVAN__ ? WrapperConstants.PROVIDER_INFURA_KOVAN : WrapperConstants.PROVIDER_INFURA_MAIN).replace('/v3', ''),
	'',
	!__KOVAN__
);
export const dualClassWrappers: ICustodianWrappers = {};
for (const type in web3Wrapper.contractAddresses.Custodians) {
	dualClassWrappers[type] = {};
	for (const tenor in web3Wrapper.contractAddresses.Custodians[type])
		dualClassWrappers[type][tenor] = new DualClassWrapper(
			web3Wrapper,
			web3Wrapper.contractAddresses.Custodians[type][tenor].custodian.address
		);
}

export const getDualClassWrapperByTypeTenor = (type: string, tenor: string) => {
	if (dualClassWrappers[type] && dualClassWrappers[type][tenor])
		return dualClassWrappers[type][tenor];
	return dualClassWrappers[WrapperConstants.BEETHOVEN][WrapperConstants.TENOR_PPT];
};

export const getDualClassAddressByTypeTenor = (type: string, tenor: string) => {
	if (web3Wrapper.contractAddresses.Custodians[type][tenor])
		return web3Wrapper.contractAddresses.Custodians[type][tenor];
	return web3Wrapper.contractAddresses.Custodians[WrapperConstants.BEETHOVEN][
		WrapperConstants.TENOR_PPT
	];
};

export const magiWrapper = new MagiWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Oracles[0].address
);

export const esplanadeWrapper = new EsplanadeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.MultiSigManagers[0].address
);

export const stakeWrapper0 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[0].address
);

export const stakeWrapper1 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[1].address
);

export const stakeWrappers = [stakeWrapper0, stakeWrapper1];

export const calculateNav = DualClassWrapper.calculateNav;

export const getTokenInterestOrLeverage = DualClassWrapper.getTokenInterestOrLeverage;
