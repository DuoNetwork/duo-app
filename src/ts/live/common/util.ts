import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import * as CST from './constants';
import { ICustodianPrice, IPriceStatus, ISourceData, IStatus } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public calculateNav(
		price: number,
		time: number,
		resetPrice: number,
		resetTime: number,
		alpha: number,
		beta: number,
		period: number,
		coupon: number
	) {
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + Math.floor((time - resetTime) / 1000 / period) * coupon;
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public range(start: number, end: number) {
		const subArray = Array.apply(null, { length: end }).map(Number.call, Number);
		return subArray.slice(start, end);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getCurrentPath(location: any) {
		return location.pathname === '/' ? location.pathname + 'beethoven' : location.pathname;
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<ICustodianPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	address: CST.DUMMY_ADDR,
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gemini: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gdax: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
