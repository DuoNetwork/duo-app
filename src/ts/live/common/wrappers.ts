import BeethovenWapper from '../../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../../duo-contract-wrapper/src/Web3Wrapper';
import infura from '../../keys/infura.json';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
