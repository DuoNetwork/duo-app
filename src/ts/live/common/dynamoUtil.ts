import AWS from 'aws-sdk';
import { ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => {
				if (err) reject(err);
				else resolve(data);
			});
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output =  status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else if (process.startsWith('FEED'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_HISTORY_PRICE].N),
					volume: Number(d[CST.DB_HISTORY_VOLUME].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
