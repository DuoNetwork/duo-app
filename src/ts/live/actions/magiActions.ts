import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import * as CST from '../common/constants';
import { IAcceptedPrice, VoidThunkAction } from '../common/types';
import util from '../common/util';

export function acceptedPricesUpdate(acceptedPrices: IAcceptedPrice[]) {
	return {
		type: CST.AC_MAG_ACCEPTED_PX,
		value: acceptedPrices
	};
}

export function fetchAcceptedPrices(contractAddress: string): VoidThunkAction {
	return async dispatch =>
		dispatch(
			acceptedPricesUpdate(
				(await dynamoUtil.queryAcceptPriceEvent(
					contractAddress,
					util.getDates(7, 1, 'day', 'YYYY-MM-DD')
				)).sort((a, b) => - a.timestamp + b.timestamp)
			)
		);
}

export function subscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_MAG_SUB,
		value: intervalId
	};
}

export function refresh(contractAddress: string): VoidThunkAction {
	return async dispatch => dispatch(fetchAcceptedPrices(contractAddress));
}

export function subscribe(contractAddress: string): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refresh(contractAddress));
		dispatch(
			subscriptionUpdate(
				window.setInterval(() => dispatch(refresh(contractAddress)), 1800000)
			)
		);
	};
}
