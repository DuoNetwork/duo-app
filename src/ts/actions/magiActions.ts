import { IMagiStates } from '@finbook/duo-contract-wrapper';
import { IAcceptedPrice } from '@finbook/duo-market-data';
import * as CST from 'ts/common/constants';
import dynamoUtil from 'ts/common/dynamoUtil';
import { VoidThunkAction } from 'ts/common/types';
import util from 'ts/common/util';
import { magiWrapper } from 'ts/common/wrappers';

export function acceptedPricesUpdate(acceptedPrices: IAcceptedPrice[]) {
	return {
		type: CST.AC_MAG_ACCEPTED_PX,
		value: acceptedPrices
	};
}

export function statesUpdate(states: IMagiStates) {
	return {
		type: CST.AC_MAG_STATES,
		value: states
	};
}
export function priceFeedUpdate(address: string, balance: number, index: number) {
	return {
		type: CST.AC_MAG_PF,
		balance: balance,
		address: address,
		index: index
	};
}

export function operatorUpdate(account: { address: string; balance: number }) {
	return {
		type: CST.AC_MAG_OPT,
		value: account
	};
}

export function roleManagerUpdate(account: { address: string; balance: number }) {
	return {
		type: CST.AC_MAG_ROLE_MNG,
		value: account
	};
}

export function fetchAcceptedPrices(contractAddress: string): VoidThunkAction {
	return async dispatch =>
		dispatch(
			acceptedPricesUpdate(
				(await dynamoUtil.queryAcceptPriceEvent(
					contractAddress,
					util.getDates(7, 1, 'day', 'YYYY-MM-DD')
				)).sort((a, b) => -a.timestamp + b.timestamp)
			)
		);
}

export function subscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_MAG_SUB,
		value: intervalId
	};
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		const addrs = await magiWrapper.getAddresses();
		dispatch(statesUpdate(await magiWrapper.getStates()));
		dispatch(fetchAcceptedPrices(magiWrapper.address));
		magiWrapper.web3Wrapper.getEthBalance(addrs.operator).then(optBalance =>
			dispatch(
				operatorUpdate({
					address: addrs.operator,
					balance: optBalance
				})
			)
		);

		magiWrapper.web3Wrapper.getEthBalance(addrs.roleManagerAddress).then(roleManagerBalance =>
			dispatch(
				roleManagerUpdate({
					address: addrs.operator,
					balance: roleManagerBalance
				})
			)
		);

		addrs.priceFeed.map((pf, i) => {
			magiWrapper.web3Wrapper
				.getEthBalance(pf)
				.then(pfBalance => dispatch(priceFeedUpdate(pf, pfBalance, i)));
		});
	};
}

export function subscribe(): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refresh());
		dispatch(subscriptionUpdate(window.setInterval(() => dispatch(refresh()), 60000)));
	};
}
