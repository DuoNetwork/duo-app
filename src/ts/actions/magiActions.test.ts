// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import util from 'ts/common/util';
import { magiWrapper } from 'ts/common/wrappers';
import dynamoUtil from '../../../../duo-admin/src/utils/dynamoUtil';
import * as magiActions from './magiActions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('acceptedPricesUpdate', () => {
		expect(magiActions.acceptedPricesUpdate(['test'] as any)).toMatchSnapshot();
	});

	test('fetchAcceptedPrices', () => {
		const store: any = mockStore({});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		store.dispatch(magiActions.fetchAcceptedPrices(WrapperConstants.DUMMY_ADDR));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('refresh', () => {
		util.getDates = jest.fn(() => ['1970-01-15']);
		const store: any = mockStore({
			magi: {
				states: {
				},
				addresses: []
			}
		});

		magiWrapper.getStates = jest.fn(() =>
			Promise.resolve({
				isStarted: false,
				firstPrice: {
					price: 100,
					timestamp: 1234567890000,
					source: 'source1'
				},
				secondPrice: {
					price: 102,
					timestamp: 1234567890000,
					source: 'source2'
				},
				priceTolerance: 0.1,
				priceFeedTolerance: 0.5,
				priceFeedTimeTolerance: 100,
				priceUpdateCoolDown: 1000,
				numOfPrices: 1
			})
		);
		magiWrapper.getAddresses = jest.fn(() =>
			Promise.resolve(
				['pf1', 'pf2', 'pf3']
			)
		);
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		store.dispatch(magiActions.refresh());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('subscriptionUpdate', () => {
		expect(magiActions.subscriptionUpdate(123)).toMatchSnapshot();
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		const store: any = mockStore({});
		dynamoUtil.queryAcceptPriceEvent = jest.fn(() => Promise.resolve(['test']));
		store.dispatch(magiActions.subscribe());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
