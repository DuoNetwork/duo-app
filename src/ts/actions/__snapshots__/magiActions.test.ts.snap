// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions acceptedPricesUpdate 1`] = `
Object {
  "type": "magiAcceptedPrices",
  "value": Array [
    "test",
  ],
}
`;

exports[`actions fetchAcceptedPrices 1`] = `
Array [
  Object {
    "type": "magiAcceptedPrices",
    "value": Array [
      Object {
        "price": 100,
        "timestamp": 1234567890000,
      },
      Object {
        "price": 102,
        "timestamp": 1234567880000,
      },
    ],
  },
]
`;

exports[`actions refresh 1`] = `
Array [
  Object {
    "type": "magiStates",
    "value": Object {
      "firstPrice": Object {
        "price": 100,
        "source": "source1",
        "timestamp": 1234567890000,
      },
      "isStarted": false,
      "numOfPrices": 1,
      "priceFeedTimeTolerance": 100,
      "priceFeedTolerance": 0.5,
      "priceTolerance": 0.1,
      "priceUpdateCoolDown": 1000,
      "secondPrice": Object {
        "price": 102,
        "source": "source2",
        "timestamp": 1234567890000,
      },
    },
  },
  Object {
    "type": "magiAcceptedPrices",
    "value": Array [
      Object {
        "price": 100,
        "timestamp": 1234567890000,
      },
      Object {
        "price": 102,
        "timestamp": 1234567880000,
      },
    ],
  },
]
`;

exports[`actions subscribe 1`] = `
Array [
  Object {
    "type": "magiSubscription",
    "value": 0,
  },
  Object {
    "type": "magiSubscription",
    "value": 123,
  },
  Object {
    "type": "magiStates",
    "value": Object {
      "firstPrice": Object {
        "price": 100,
        "source": "source1",
        "timestamp": 1234567890000,
      },
      "isStarted": false,
      "numOfPrices": 1,
      "priceFeedTimeTolerance": 100,
      "priceFeedTolerance": 0.5,
      "priceTolerance": 0.1,
      "priceUpdateCoolDown": 1000,
      "secondPrice": Object {
        "price": 102,
        "source": "source2",
        "timestamp": 1234567890000,
      },
    },
  },
  Object {
    "type": "magiAcceptedPrices",
    "value": Array [
      Object {
        "price": 100,
        "timestamp": 1234567890000,
      },
      Object {
        "price": 102,
        "timestamp": 1234567880000,
      },
      Object {
        "price": 106,
        "timestamp": 1234567870000,
      },
    ],
  },
]
`;

exports[`actions subscribe 2`] = `60000`;

exports[`actions subscribe 3`] = `
Array [
  Object {
    "type": "magiSubscription",
    "value": 0,
  },
  Object {
    "type": "magiSubscription",
    "value": 123,
  },
  Object {
    "type": "magiStates",
    "value": Object {
      "firstPrice": Object {
        "price": 100,
        "source": "source1",
        "timestamp": 1234567890000,
      },
      "isStarted": false,
      "numOfPrices": 1,
      "priceFeedTimeTolerance": 100,
      "priceFeedTolerance": 0.5,
      "priceTolerance": 0.1,
      "priceUpdateCoolDown": 1000,
      "secondPrice": Object {
        "price": 102,
        "source": "source2",
        "timestamp": 1234567890000,
      },
    },
  },
  Object {
    "type": "magiAcceptedPrices",
    "value": Array [
      Object {
        "price": 100,
        "timestamp": 1234567890000,
      },
      Object {
        "price": 102,
        "timestamp": 1234567880000,
      },
      Object {
        "price": 106,
        "timestamp": 1234567870000,
      },
    ],
  },
]
`;

exports[`actions subscriptionUpdate 1`] = `
Object {
  "type": "magiSubscription",
  "value": 123,
}
`;
